lisa/sut_orchestrator/hyperv/platform_.py

```diff
@@ -334,3 +378,255 @@ def _delete_node(node_ctx: NodeContext, wait_delete: bool) -> None:
+    def __init__(
+        self,
+        node: RemoteNode,
+        assignable_device_script: PurePath,
+        runbook: HypervPlatformSchema,
+    ) -> None:
+        # Device Passthrough configs
+        # Mapping of Host Device Passthrough
+        self.AVAILABLE_HOST_DEVICES: Dict[
+            HostDevicePoolType, List[DeviceAddressSchema]
+        ] = {}
+        self.SUPPORTED_HOST_DEVICE_POOLTYPE = [
+            HostDevicePoolType.PCI_NIC,
+            HostDevicePoolType.PCI_GPU,
+        ]
+        self.POOL_TYPE_TO_DEVICE_PROPERTY = {
+            HostDevicePoolType.PCI_NIC: DeviceAddressSchema,
+            HostDevicePoolType.PCI_GPU: DeviceAddressSchema,
+        }
+        self._server = node
+        self._assignable_devices_script = assignable_device_script
+        self._hyperv_runbook = runbook
+
+    def _create_device_pool(
+        self,
+        pool_type: HostDevicePoolType,
+        vendor_id: str,
+        device_id: str,
+    ) -> None:
+        self.AVAILABLE_HOST_DEVICES[pool_type] = []
+        powershell = self._server.tools[PowerShell]
+        cmdlet = (
+            f"{self._assignable_devices_script} "
+            f"-vendorId {vendor_id} -deviceId {device_id}"
+        )
+        stdout = powershell.run_cmdlet(
+            cmdlet=cmdlet,
+            sudo=True,
+            force_run=True,
+        ).strip()
+
+        if stdout.find("The list is empty") >= 0:
+            return
+        else:
+            match = re.search(
+                r"Assignable Devices Found:(.*)",
+                stdout,
+                re.DOTALL,
+            )
+            if match:
+                devices_csv_list = match.group(1).strip()
+                primary_nic_id_list = self._get_primary_nic_id()
+                assert len(primary_nic_id_list) != 0
+                for device_record in devices_csv_list.splitlines():
+                    device_attrbs = device_record.strip().split(",")
+                    assert len(device_attrbs) == 3
+                    instance_id = device_attrbs[2].replace('"', "")
+                    if instance_id not in primary_nic_id_list:
+                        device = self.POOL_TYPE_TO_DEVICE_PROPERTY[pool_type]()
+                        device.friendly_name = device_attrbs[0].replace('"', "")
+                        device.location_path = device_attrbs[1].replace('"', "")
+                        device.instance_id = instance_id
+
+                        # add device to the given pool_type
+                        pool = self.AVAILABLE_HOST_DEVICES.get(pool_type, [])
+                        pool.append(device)
+                        self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+
+    def _get_devices_from_pool(
+        self,
+        pool_type: HostDevicePoolType,
+        count: int,
+    ) -> List[DeviceAddressSchema]:
+        pool = self.AVAILABLE_HOST_DEVICES[pool_type]
+        if len(pool) < count:
+            raise SkippedException(
```

Comment:
It should raise exceptions, so the runbook and test cases should be changed to avoid test case cannot run, because of lack of resources. You can also raise `ResourceAwaitableException`, the lisa runner will follow the retry mechanism, and it may be assigned next time.

Reply:
Sure Chi, will use `ResourceAwaitableException`

Reply:
@squirrelsc Please check the latest push